File: 백준/Gold/5427. 불/불.py
import sys
from collections import deque

input = sys.stdin.readline

dx = [0, 0, 1, -1]
dy = [1, -1, 0, 0]

def bfs():
    global time

    while start:
        time += 1
        for j in range(len(fire)):
            x, y = fire.popleft()

            for i in range(4):
                nx = x + dx[i]
                ny = y + dy[i]

                if 0 <= nx < h and 0 <= ny < w and maps[nx][ny] == ".":
                    maps[nx][ny] = "*"
                    fire.append((nx, ny))
        for j in range(len(start)):
            x, y = start.popleft()

            for i in range(4):
                nx = x + dx[i]
                ny = y + dy[i]

                if 0 <= nx < h and 0 <= ny < w:
                    if maps[nx][ny] == ".":
                        maps[nx][ny] = "@"
                        start.append((nx, ny))
                else:
                    return time + 1
    return "IMPOSSIBLE"
for _ in range(int(input())):
    w, h = map(int, input().split())
    time = -1
    maps = []
    fire = deque()
    start = deque()

    for i in range(h):
        row = list(map(str, input().rstrip()))
        for j in range(w):
            if row[j] == "@":
                start.append((i, j))
            elif row[j] == "*":
                fire.append((i, j))
        maps.append(row)
    print(bfs())

File: 백준/Gold/5427. 불/불.java
import java.io.*;
import java.util.*;

class Main {
    static int h, w;
    static Character [][] maps;
    static int [][] visit;
    static int dx [] = {0, 0, 1, -1};
    static int dy [] = {1, -1, 0, 0};
    static Queue<Point> Fire = new LinkedList<>();

    public static class Point {
        int x, y, c;
        public Point(int x, int y) {
            this.x = x;
            this.y = y;
        }
        public Point(int x, int y, int c) {
            this.x = x;
            this.y = y;
            this.c = c;
        }
    }

    public static int BFS(Point start) {
        Queue<Point> queue = new LinkedList<Point>();
        visit[start.x][start.y] = 0;
        
        queue.offer(start);
        while (!queue.isEmpty()) {
            
            int firesize = Fire.size();
            for (int fs = 0; fs < firesize; fs++) {
                Point fcurrent = Fire.poll();

                for (int i = 0; i < 4; i++) {
                    int nx = fcurrent.x + dx[i];
                    int ny = fcurrent.y + dy[i];

                    if (0 <= nx && nx < h && 0 <= ny && ny < w && maps[nx][ny] == '.') {
                        maps[nx][ny] = '*';
                        
                        Fire.offer(new Point(nx, ny));
                    }
                }
            }

            int qsize = queue.size();
            for (int s = 0; s < qsize; s++) {
                Point current = queue.poll();

                for (int i = 0; i < 4; i++) {
                    int nx = current.x + dx[i];
                    int ny = current.y + dy[i];

                    if (0 <= nx && nx < h && 0 <= ny && ny < w){
                        if (maps[nx][ny] == '.') {
                            maps[nx][ny] = '@';
                            visit[nx][ny] = visit[current.x][current.y] + 1;
                            queue.offer(new Point(nx, ny));
                        }
                    }
                    else {
                        return visit[current.x][current.y] + 1;
                    }
                }
            }
        }
        return -1;
    }
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine());

        for(int k = 0; k < n; k++) {
            String [] nums = br.readLine().split(" ");

            h = Integer.parseInt(nums[1]);
            w = Integer.parseInt(nums[0]);

            maps = new Character [h][w];
            visit = new int [h][w];
            Fire = new LinkedList<Point>();

            int s_x = 0;
            int s_y = 0;
            for (int i = 0; i < h; i++) {
                String row = br.readLine();
                for (int j = 0; j < w; j++) {
                    maps[i][j] = row.charAt(j);
                    if (maps[i][j] == '@') {
                        s_x = i;
                        s_y = j;
                    }
                    if (maps[i][j] == '*') {
                        Fire.offer(new Point(i, j));
                    }
                }
            }
            int ans = BFS(new Point(s_x, s_y, 0));
            if (ans == -1) System.out.println("IMPOSSIBLE");
            else System.out.println(ans);
        }
    }
}